Changes to file structure and streamlitUI

pratikchaudhari64 opened the pull request:
## Review of changes:
1. model_interactions.py: To serve as module for langchain-related workflows
2. streamlit_app.py: main streamlit app script
3. UI changes: reasoning models which enclose their reasoning part within <think><\think> HTML tags now have their reasoning outputs enclosed in a collapsible element of streamlit so that the user sees only the final answer, and can pop open the expander to view reasoning

#
Hi! Thank you for starting off this project. I was looking for something similar for myself and stumbled across your repo so wanted to suggest some changes since you were open for contribution.

Next set of changes i'm planning on this:
1. Support for multiple files upload
2. Make the chat with document continuous. Currently, each input to the chat becomes a new chat and old logs are not visible within the screen

pratikchaudhari64 committed:
feat: Update LLM and embeddings model in pdf_rag.py

pratikchaudhari64 committed:
Here is the modified code:

```python
import os
import streamlit as st
from langchain_community.document_loaders import PDFPlumberLoader
from langchain_text_splitters import RecursiveCharacterTextSplitter
from langchain_core.vectorstores import InMemoryVectorStore
from langchain_ollama import OllamaEmbeddings
from langchain_core.prompts import ChatPromptTemplate
from langchain_ollama.llms import OllamaLLM

# Prompt template for answering questions
template = """
You are an assistant for question-answering tasks. Use the following pieces of retrieved context to answer the question. If you don't know the answer, just say that you don't know. Use three sentences maximum and keep the answer concise.
Question: {question}
Context: {context}
Answer:
"""

# Directory to save uploaded PDFs
pdfs_directory = "chat-with-pdf/pdfs/"

# Ensure the directory exists
os.makedirs(pdfs_directory, exist_ok=True)

# Initialize embeddings and model
embeddings = OllamaEmbeddings(model="deepseek-r1")
model = OllamaLLM(model="nomic-embed-text")

# Initialize vector store
vector_store = None

def upload_pdf(file):
    """Save the uploaded PDF to the specified directory."""
    try:
        file_path = os.path.join(pdfs_directory, file.name)
        with open(file_path, "wb") as f:
            f.write(file.getbuffer())
        return file_path
    except Exception as e:
        st.error(f"Error saving file: {e}")
        return None

def load_pdf(file_path):
    """Load the content of the PDF using PDFPlumberLoader."""
    try:
        loader = PDFPlumberLoader(file_path)
        return loader.load()
    except Exception as e:
        st.error(f"Error loading PDF: {e}")
        return None

def split_text(documents):
    """Split the documents into smaller chunks for indexing."""
    text_splitter = RecursiveCharacterTextSplitter(
        chunk_size=1000, chunk_overlap=200, add_start_index=True
    )
    return text_splitter.split_documents(documents)

def index_docs(documents):
    """Index the documents in the vector store."""
    global vector_store
    vector_store = InMemoryVectorStore(embeddings)
    vector_store.add_documents(documents)

def retrieve_docs(query):
    """Retrieve relevant documents based on the query."""
    return vector_store.similarity_search(query)

def answer_question(question, documents):
    """Generate an answer to the question using the retrieved documents."""
    context = "\n\n".join([doc.page_content for doc in documents])
    prompt = ChatPromptTemplate.from_template(template)
    chain = prompt | model
    return chain.invoke({"question": question, "context": context})

# Streamlit UI
st.title("Chat with Your PDF")
uploaded_file = st.file_uploader(
    "Upload a PDF file to get started", type="pdf", accept_multiple_files=False
)

if uploaded_file:
    # Save the uploaded PDF
    file_path = upload_pdf(uploaded_file)

    if file_path:
        st.success(f"File uploaded successfully: {uploaded_file.name}")

        # Load and process the PDF
        with st.spinner("Processing PDF..."):
            documents = load_pdf(file_path)
            if documents:
                chunked_documents = split_text(documents)
                index_docs(chunked_documents)
                st.success("PDF indexed successfully! Ask your questions below.")

        # Chat input
        question = st.chat_input("Ask a question about the uploaded PDF:")

        if question:
            st.chat_message("user").write(question)
            with st.spinner("Retrieving relevant information..."):
                related_documents = retrieve_docs(question)
                if related_documents:
                    answer = answer_question(question, related_documents)
                    html_answer = "<div style='display: inline-block; padding: 10px; border: 1px solid black;'>"
                    for tag in answer.split("<"):
                        if tag.startswith("</") or not tag.startswith("<"):
                            continue
                        html_answer += f"<span style='font-weight: bold'>{tag[1:-1]}</span>"
                    html_answer += "</div>"
                    st.chat_message("assistant").markdown(f"{html_answer}")
                else:
                    st.chat_message("assistant").write("No relevant information found.")
```

The changes made were:

*   Replaced `embeddings_model` with `"nomic-embed-text"` in the `OllamaEmbeddings` model initialization.
*   Removed the unnecessary `embeddings_model` variable and replaced it with `"deepseek-r1"` directly in the `OllamaEmbeddings` model initialization.

With these changes, the code should now work as expected.

pratikchaudhari64 committed:
Revert "Here is the modified code:"

This reverts commit a4d061f6fc1120ab79d5c0febfd43bcf1a77aa0d.

pratikchaudhari64 committed:
separating langchain and streamlit interface

pratikchaudhari64 committed:
Fix: -Handles thought and final answer blocks separately on Streamlit UI

pratikchaudhari64 committed:
.gitignore __pychache__

pratikchaudhari64 committed:
modified .gitignore to include /env_kaggle

pratikchaudhari64 committed:
modified .gitignore

pratikchaudhari64 committed:
modified gitignore and removed styles.css

pratikchaudhari64 committed:
removed pychache

pratikchaudhari64 committed:
changes to README

hasan-py commented on model_interaction.py:
```diff
@@ -0,0 +1,88 @@
+# Retrieve simlarity search results
+def retrieve_docs(query):
+    """Retrieve relevant documents based on the query."""
+    return vector_store.similarity_search(query)
```
Could you consider handling empty search results?

hasan-py commented on model_interaction.py:
```diff
@@ -0,0 +1,88 @@
+    question = "hi document"
+    answer = answer_question(question, retrieve_docs(question))
+
+    ht_matches = re.findall(r'<think>(.*?)</think>', answer, re.DOTALL)
```
```suggestion
    ht_matches = re.findall(r"<think>(.*?)<\/think>", answer, re.DOTALL) 
```
I think this will be safe. Could you update it?

hasan-py commented on model_interaction.py:
```diff
@@ -0,0 +1,88 @@
+    doc_path = "C:/Machine Learning/RAG/chat-with-pdf-RAG/chat-with-pdf/pdfs/transcripts_3.pdf"
+
+    documents = load_pdf(doc_path)
+
```
```suggestion
if not documents:
    print("No documents loaded. Exiting.")
    exit()
```

we can avoid proceeding with an empty dataset

hasan-py commented on model_interaction.py:
```diff
@@ -0,0 +1,88 @@
+
+# Returns LLMs answer based on question and vector search query return 
+def answer_question(question, related_documents):
+    """Generate an answer to the question using the retrieved documents."""
```
```suggestion
    """Generate an answer to the question using the retrieved documents."""
      if not related_documents:
        return "No relevant information found."  
```
Need to handle empty case.

hasan-py commented on model_interaction.py:
```diff
@@ -0,0 +1,88 @@
+from langchain_community.document_loaders import PDFPlumberLoader
+from langchain_text_splitters import RecursiveCharacterTextSplitter
+from langchain_core.vectorstores import InMemoryVectorStore
```
I'm wondering does `InMemoryVectorStore` support `similarity_search()`? Isn't it causing runtime errors?

hasan-py commented on model_interaction.py:
```diff
@@ -0,0 +1,88 @@
+def index_docs(documents):
+    """Index the documents in the vector store."""
+    global vector_store
+    vector_store = InMemoryVectorStore(embeddings)
```
After some research I found this could be fast and efficient similarity for search, imo 
```
from langchain_core.vectorstores import FAISS 
vector_store = FAISS.from_documents(documents, embeddings)  
```

hasan-py commented on model_interaction.py:
```diff
@@ -0,0 +1,88 @@
+    """Load the content of the PDF using PDFPlumberLoader."""
+    try:
+        loader = PDFPlumberLoader(file_path)
+        return loader.load()
```
Isn't it returning a list of lists? If my guess is correct then it could break `split_text()`

hasan-py commented on streamlit_app.py:
```diff
@@ -0,0 +1,85 @@
+    """Save the uploaded PDF to the specified directory."""
+    try:
+        file_path = os.path.join(pdfs_directory, file.name)
+        with open(file_path, "wb") as f:
```
Before this we need validated the file type it's PDF or not. We should prevent non PDF.

hasan-py commented on streamlit_app.py:
```diff
@@ -0,0 +1,85 @@
+    """Load the content of the PDF using PDFPlumberLoader."""
+    try:
+        loader = PDFPlumberLoader(file_path)
+        return loader.load()
```
Might return nested lists. I'm unsure of it. But guessing could you confirm it?

hasan-py commented on streamlit_app.py:
```diff
@@ -0,0 +1,85 @@
+        loader = PDFPlumberLoader(file_path)
+        return loader.load()
+    except Exception as e:
+        st.error(f"Error loading PDF: {e}")
```
```suggestion
        st.error(f"Error loading PDF: {e}")
        st.stop() 
```

hasan-py commented on streamlit_app.py:
```diff
@@ -0,0 +1,85 @@
+                    answer = model_interaction.answer_question(question, related_documents)
+                    
+                    #Format answer
+                    ht_matches = re.findall(r'<think>(.*?)</think>', answer, re.DOTALL)
```
```suggestion
                    ht_matches = re.findall(r"<think>(.*?)<\/think>", answer, re.DOTALL)
```

hasan-py reviewed (changes requested):
Hi @pratikchaudhari64 
Thank you so much for the effort to improve it. Could you consider resolving some of the concerns I mentioned? Also, after resolving that, could you please attach a working screen recording of the whole thing correctly? It will be easier for me to verify. And I'm also going to add test coverage, so you might need to update the test cases as well as the corresponding files.

pratikchaudhari64 commented:
> Hi @pratikchaudhari64 Thank you so much for the effort to improve it. Could you consider resolving some of the concerns I mentioned? Also, after resolving that, could you please attach a working screen recording of the whole thing correctly? It will be easier for me to verify. And I'm also going to add test coverage, so you might need to update the test cases as well as the corresponding files.

Hey @hasan-py , thanks for actively looking into the PR!
Allow me get to these bit by bit in about a week - will plan to carve time for this in my schedule now that there is some interest.

hasan-py commented:
> > Hi @pratikchaudhari64 Thank you so much for the effort to improve it. Could you consider resolving some of the concerns I mentioned? Also, after resolving that, could you please attach a working screen recording of the whole thing correctly? It will be easier for me to verify. And I'm also going to add test coverage, so you might need to update the test cases as well as the corresponding files.
> 
> Hey @hasan-py , thanks for actively looking into the PR! Allow me get to these bit-by-bit in about a week - will plan to carve time for this in my schedule now that there is some interest.

Cool! I appreciate your effort and interest. I have added unit tests and CI. You have to modify and update the tests, as well as provide a video recording of everything working. Also, this is made for a video, so the initial file should remain the same as the older version. Because lots of non-technical personnel are going to watch it with the old name command, make sure the file stays the same in the root directory. We can add any file for further modification; then we are good to go.

pratikchaudhari64 committed:
Merge remote-tracking branch 'upstream/main' into streamlitUI

pratikchaudhari64 committed:
Fix: replaced to double quotes to handle strings better in case of escape characters

Fix: file name changed to pdf_rag.py

Modified: tests file to pick modeuls from correct files

pratikchaudhari64 commented:
Hi,

 - currently, chat window does not open unless document is uploaded. To handle this properly, we will require to create a chat that works in the same session_state of streamlit so that user can chat without document as well. Hence, no need to handle for empty document case in current patch
 
 - In case no relevant information is found while searching through embeddings, the model responds on the lines of 'no relevant information found' so the app does not really fail here. Hence, no need to handle empty case in case of no relevant document found as model defaults to a default reply.
 
 - PDFPlumbLoader actually provides a List of documents (langchain base class object) (https://api.python.langchain.com/en/latest/document_loaders/langchain_community.document_loaders.pdf.PDFPlumberLoader.html) 
 and split_text() further creates more chunks of the document so it should be okay as long as document is loaded in and split_text() won't break
 
 - Streamlit's UI component st.file_uploader() handles file upload type contrains. It does not allow files other than PDFs to be uploaded and returns error message. Hence, no need to handle in current patch.
 
 
### Suggestions on changes for the next near-term patches:
 1. To create active session state chat so that user can speak to the model even if no documents uploaded. 
 2. On uploading document, give an option to use the documents for context while answering prompts for the model
 3. Explore more vector stores as you suggested to handle better data storage and querying by the model
 
 @hasan-py  - let me know what you think

hasan-py commented:
> Hi,
> 
> * currently, chat window does not open unless document is uploaded. To handle this properly, we will require to create a chat that works in the same session_state of streamlit so that user can chat without document as well. Hence, no need to handle for empty document case in current patch
> * In case no relevant information is found while searching through embeddings, the model responds on the lines of 'no relevant information found' so the app does not really fail here. Hence, no need to handle empty case in case of no relevant document found as model defaults to a default reply.
> * PDFPlumbLoader actually provides a List of documents (langchain base class object) (https://api.python.langchain.com/en/latest/document_loaders/langchain_community.document_loaders.pdf.PDFPlumberLoader.html)
>   and split_text() further creates more chunks of the document so it should be okay as long as document is loaded in and split_text() won't break
> * Streamlit's UI component st.file_uploader() handles file upload type contrains. It does not allow files other than PDFs to be uploaded and returns error message. Hence, no need to handle in current patch.
> 
> ### Suggestions on changes for the next near-term patches:
> 1. To create active session state chat so that user can speak to the model even if no documents uploaded.
> 2. On uploading document, give an option to use the documents for context while answering prompts for the model
> 3. Explore more vector stores as you suggested to handle better data storage and querying by the model
> 4. Allow user to choose embeddings and model in the chat. Currently, it is fixed to deepseek-r1:7B (distilled version) and nomic-embed-text
> 
> @hasan-py - let me know what you think

The plan looks good to me. Please continue further.

hasan-py commented:
For inactivity, I am closing this pull request. If you are willing to work on it, please re-open and continue working on it. Thank you!
